torchmultimodal/models/vqvae.py

```diff
@@ -52,18 +53,30 @@ def __init__(
         self.decoder = decoder
         self.codebook = Codebook(codebook_num_embeddings, codebook_embedding_dim)
 
-    def encode(self, x: Tensor) -> CodebookOutput:
-        return self.codebook(self.encoder(x))
+    def latent_shape(self, input_shape: Union[Size, Tuple]) -> Tuple[int, ...]:
+        return self.encoder.get_latent_shape(input_shape)  # type: ignore
+
+    def encode(
+        self, x: Tensor, return_embeddings: bool = False
+    ) -> Union[Tuple[Tensor, Tensor], Tensor]:
+        encoded = self.encoder(x)
+        out = self.codebook(encoded)
+        indices = out.codebook_indices
+        quantized = out.quantized
+        if return_embeddings:
+            return indices, quantized
+        return indices
 
-    def decode(self, x: Tensor) -> Tensor:
-        return self.decoder(x)
+    def decode(self, indices: Tensor) -> Tensor:
+        quantized = self.codebook.lookup(indices)  # (b, latent_shape, emb_dim)
+        quantized = shift_dim(quantized, -1, 1)  # (b, emb_dim, latent_shape)
+        return self.decoder(quantized)  # (b, c, input_shape)
 
-    def tokenize(self, x: Tensor) -> Tensor:
-        """Similar to encode, but return flattened quantized outputs"""
-        quantized = self.encode(x)
-        return quantized.quantized_flat
+    def lookup(self, indices: Tensor) -> Tensor:
```

Comment:
maybe should check that codebook module has `lookup` method and throw `AttributeError` if it doesn't

Reply:
Added